src/main/java/network/brightspots/rcv/CommonDataFormatReader.java

```diff
@@ -336,51 +274,214 @@ void parseCvrFile(List<CastVoteRecord> castVoteRecords)
+        String candidateId;
+        if (contestSelection.containsKey("IsWriteIn") && contestSelection.get("IsWriteIn")
+            .equals(BOOLEAN_TRUE)) {
+          // this is a write-in
+          candidateId = this.config.getUndeclaredWriteInLabel();
+          if (candidateId == null || candidateId.isEmpty()) {
+            Logger.severe(
+                "Undeclared write-in candidate found while parsing CVR, however undeclared "
+                    + "write-in label has not been defined in config.");
+            throw new CvrParseException();
+          }
+        } else {
+          // lookup Candidate Name
+          ArrayList candidateIds = (ArrayList) contestSelection.get("CandidateIds");
+          if (candidateIds == null || candidateIds.size() == 0) {
+            Logger.severe("CandidateSelection \"%s\" has no CandidateIds!", contestSelectionId);
+            throw new CvrParseException();
+          }
+          if (candidateIds.size() > 1) {
+            Logger.warning("CandidateSelection \"%s\" has multiple CandidateIds.  "
+                + "Only the first one will be processed.", contestSelectionId);
+          }
+          String candidateObjectId = (String) candidateIds.get(0);
+          HashMap candidate = (HashMap) candidates.get(candidateObjectId);
+          candidateId = (String) candidate.get("Name");
+          if (candidateId.equals(config.getOvervoteLabel())) {
+            candidateId = Tabulator.EXPLICIT_OVERVOTE_LABEL;
+          } else if (!this.config.getCandidateCodeList().contains(candidateId)) {
+            Logger.severe("Unrecognized candidate found in CVR: %s", candidateId);
+            unrecognizedCandidateCounts.merge(candidateId, 1, Integer::sum);
+          }
+        }
+
+        // parse rankings
+        // rank may appear on the CVRContestSelection OR the SelectionPosition
+        // this is an ambiguity in the nist spec
+        if (cvrContestSelection.containsKey("Rank")) {
+          Integer rank = (Integer) (cvrContestSelection.get("Rank"));
+          rankings.add(new Pair<>(rank, candidateId));
+        } else {
+          // extract all the SelectionPositions (ranks) which this selection has been assigned
+          ArrayList selectionPositions = (ArrayList) cvrContestSelection.get("SelectionPosition");
+          for (Object selectionPositionObject : selectionPositions) {
+            HashMap selectionPosition = (HashMap) selectionPositionObject;
+            // WriteIn can be linked at the selection position level
+            if (selectionPosition.containsKey("CVRWriteIn")) {
+              candidateId = this.config.getUndeclaredWriteInLabel();
+              if (candidateId == null || candidateId.isEmpty()) {
+                Logger.severe(
+                    "Undeclared write-in candidate found while parsing CVR, however undeclared "
+                        + "write-in label has not been defined in config.");
+                throw new CvrParseException();
+              }
```

Comment:
@moldover I added an inline comment here but GitHub's crappy UI seems to have discarded it...? Anyway, the comment was that you added the same six lines in four different places, so please factor it out into a helper. Also, you can use `isNullOrBlank`.